event.go

```diff
@@ -0,0 +1,178 @@
+package mesg
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	"github.com/mesg-foundation/core/api/core"
+)
+
+// Event is a MESG event.
+type Event struct {
+	Key  string
+	data string
+}
+
+// Data decodes event data into out.
+func (e *Event) Data(out interface{}) error {
+	return json.Unmarshal([]byte(e.data), out)
+}
+
+// EventEmitter is a MESG event emitter.
+type EventEmitter struct {
+	app *Application
+
+	// eventKey is the actual event to listen for.
+	eventKey string
+
+	//eventServiceID is the service id of where event is emitted.
+	eventServiceID string
+
+	// taskKey is the actual task that will be executed.
+	taskKey string
+
+	// taskServiceID is the service id of target task.
+	taskServiceID string
+
+	// filterFuncs holds funcs that returns boolean values to decide
+	// if the task should be executed or not.
+	filterFuncs []func(*Event) bool
+
+	// mapFunc is a func that returns input data of task.
+	mapFunc func(*Event) Data
+
+	// m protects emitter configuration.
+	m sync.Mutex
+
+	gracefulWait *sync.WaitGroup
+}
+
+// EventCondition is the condition configurator for filtering events.
+type EventCondition func(*EventEmitter)
+
+// EventKeyCondition returns a new condition to filter events by name.
+// Default is all(*).
+func EventKeyCondition(event string) EventCondition {
+	return func(l *EventEmitter) {
+		l.eventKey = event
+	}
+}
+
+// WhenEvent creates an EventEmitter for serviceID.
+func (a *Application) WhenEvent(serviceID string, conditions ...EventCondition) *EventEmitter {
+	e := &EventEmitter{
+		app:            a,
+		eventServiceID: serviceID,
+		eventKey:       "*",
+		gracefulWait:   &sync.WaitGroup{},
+	}
+	for _, condition := range conditions {
+		condition(e)
+	}
+	return e
+}
+
+// Filter sets filter funcs that will be executed to decide to execute the
+// task or not.
+// It's possible to add multiple filters by calling Filter multiple times.
+// Other filter funcs and the task execution will no proceed if a filter
+// func returns false.
+func (e *EventEmitter) Filter(fn func(*Event) bool) *EventEmitter {
+	e.m.Lock()
+	defer e.m.Unlock()
+	e.filterFuncs = append(e.filterFuncs, fn)
+	return e
+}
+
+// Map sets the returned data as the input data of task.
+// You can dynamically produce input values for task over event data.
+func (e *EventEmitter) Map(fn func(*Event) Data) Executor {
+	e.m.Lock()
+	defer e.m.Unlock()
+	e.mapFunc = fn
+	return e
+}
+
+// Execute starts for listening events and executes task for serviceID with the
+// output data of event or return value of Map if all Filter funcs returned as true.
+func (e *EventEmitter) Execute(serviceID, task string) (*Listener, error) {
+	e.taskServiceID = serviceID
+	e.taskKey = task
+	listener := newListener(e.app, e.gracefulWait)
+	if err := e.app.startServices(e.eventServiceID, serviceID); err != nil {
```

Comment:
we should also start the `taskServiceID` both the event and the task services need to be running

Reply:
forget about this one, this one is ok... my bad